sympy/core/basic.py

```diff
@@ -124,6 +124,57 @@ def __new__(cls, *args):
         obj._args = args  # all items in args must be Basic objects
         return obj
 
+    def has_free(self, x, any=True, all=None):
+        """return True if self has object(s) ``x`` as a free expression
+        else False. If all objects must appear, set ``all=True``.
+
+        Examples
+        ========
+
+        >>> from sympy import Integral, Function
+        >>> from sympy.abc import x, y
+        >>> f = Function('f')
+        >>> g = Function('g')
+        >>> expr = Integral(f(x), (f(x), 1, f(y)))
+        >>> expr.free_symbols
+        {y}
+        >>> expr.has_free(f(y))
+        True
+        >>> expr.has_free(x)
+        False
+
+        This only works for expressions, not types:
+
+        >>> expr.has_free(f)
+        False
+
+        """
+        from sympy.core.symbol import Symbol
+        from sympy.utilities.iterables import iterable, sift
+        masked = self.as_dummy() if not isinstance(self, Symbol) else self
```

Comment:
If we need to use `as_dummy` then why isn't this just `return self.as_dummy().has(x)`? Any performance improvements could then be made in `has`.

Or have I misunderstood what this method does?

Reply:
I made it so for the case of `any` (i.e. `all=False`); but handling for `all=True` is also included.

Reply:
Also, the failing docstring is a result of using `has` instead of the previous code. `f` is not an expression -- does not appear as any object's `arg` so the former code that only traversed args worked. The `has` walk is more general. If we want to use `has` then I need a way to identify the generators. With #22267 `has_free` is:
```python
def has_free(self, x):
    return x in generators(self)

def all_free(self, s):
    return not (generators(self) - s)
```
But then we really only need the `generators` since this is a pretty straightforward expression.

Reply:
The problem with generators is that we have to traverse the whole tree to see if they are there, but a `has` query can exit as soon as one is hit. Building `as_dummy` must also traverse the tree, however. Perhaps an improvement would be to have `iter_args_free` methods. The Basic default would be:
```python
def iter_args_free(self):
    for a in self.args:
        yield a
        yield from a.iter_args_free
```
But the routines that have bound symbols would have their own semantics for how to do this. But all the logic is already present in the `as_dummy` construction` so there would be duplication of efforts to do the iterargs.